src/coreclr/src/vm/assemblynative.cpp

```diff
@@ -31,55 +31,52 @@
 
-FCIMPL5(Object*, AssemblyNative::Load, AssemblyNameBaseObject* assemblyNameUNSAFE,
-        AssemblyBaseObject* requestingAssemblyUNSAFE,
-        StackCrawlMark* stackMark,
-        CLR_BOOL fThrowOnFileNotFound,
-        AssemblyLoadContextBaseObject *assemblyLoadContextUNSAFE)
+/* static */
+void QCALLTYPE AssemblyNative::InternalLoad(QCall::ObjectHandleOnStack assemblyName,
+                                            QCall::ObjectHandleOnStack requestingAssembly,
+                                            QCall::StackCrawlMarkHandle stackMark,
+                                            BOOL fThrowOnFileNotFound,
+                                            QCall::ObjectHandleOnStack assemblyLoadContext,
+                                            QCall::ObjectHandleOnStack retAssembly)
 {
-    FCALL_CONTRACT;
-
-    struct _gc
-    {
-        ASSEMBLYNAMEREF        assemblyName;
-        ASSEMBLYREF            requestingAssembly;
-        ASSEMBLYREF            rv;
-        ASSEMBLYLOADCONTEXTREF assemblyLoadContext;
-    } gc;
+    QCALL_CONTRACT;
 
-    gc.assemblyName        = (ASSEMBLYNAMEREF)        assemblyNameUNSAFE;
-    gc.requestingAssembly  = (ASSEMBLYREF)            requestingAssemblyUNSAFE;
-    gc.rv                  = NULL;
-    gc.assemblyLoadContext = (ASSEMBLYLOADCONTEXTREF) assemblyLoadContextUNSAFE;
+    BEGIN_QCALL;
 
-    HELPER_METHOD_FRAME_BEGIN_RET_PROTECT(gc);
+    GCX_COOP();
 
-    if (gc.assemblyName == NULL)
+    if (*assemblyName.m_ppObject == NULL)
+    {
         COMPlusThrow(kArgumentNullException, W("ArgumentNull_AssemblyName"));
-
+    }
     ACQUIRE_STACKING_ALLOCATOR(pStackingAllocator);
 
     DomainAssembly * pParentAssembly = NULL;
     Assembly * pRefAssembly = NULL;
+    ICLRPrivBinder *pBinderContext = NULL;
 
-    INT_PTR ptrLoadContextBinder = (gc.assemblyLoadContext != NULL) ? gc.assemblyLoadContext->GetNativeAssemblyLoadContext() : NULL;
-
-    if(gc.assemblyName->GetSimpleName() == NULL)
+    if (*assemblyLoadContext.m_ppObject != NULL)
+    {
+        ASSEMBLYLOADCONTEXTREF * pAssemblyLoadContextRef = reinterpret_cast<ASSEMBLYLOADCONTEXTREF *>(assemblyLoadContext.m_ppObject);
+        INT_PTR nativeAssemblyLoadContext = (*pAssemblyLoadContextRef)->GetNativeAssemblyLoadContext();
+        pBinderContext = reinterpret_cast<ICLRPrivBinder*>(nativeAssemblyLoadContext);
+    }
+    ASSEMBLYNAMEREF * pAssemblyNameRef = reinterpret_cast<ASSEMBLYNAMEREF *>(assemblyName.m_ppObject);
+    if ((*pAssemblyNameRef)->GetSimpleName() == NULL)
     {
         COMPlusThrow(kArgumentException, W("Format_StringZeroLength"));
     }
     else
     {
         // Compute parent assembly
-        if (gc.requestingAssembly != NULL)
+        if (*requestingAssembly.m_ppObject != NULL)
         {
-            pRefAssembly = gc.requestingAssembly->GetAssembly();
+            pRefAssembly = (*(reinterpret_cast<ASSEMBLYREF *>(requestingAssembly.m_ppObject)))->GetAssembly();
```

Comment:
These casts from `m_ppObject` to `ASSEMBLYREF*`, and other similar ones, seems to be causing the asserts to fire.

It may be best to write this like this: 
```
 pRefAssembly = (ASSEMBLYREF)(requestingAssembly.Get());
```

Add `Get` method to `ObjectHandleOnStack` that looks like this:

```
OBJECTREF Get()
{
    LIMITED_METHOD_CONTRACT;
    return ObjectToOBJECTREF(*m_ppObject);
}
```

Reply:
I updated m_ppobject access to Get() except assemblyName in spec.InitializeSpec.

Reply:
Oops, GC_PROTECT for assemblyName in spec.InitializeSpec is checked.
Sorry, I will update tomorrow. Thank you.